include/pybind11/eigen.h

```diff
@@ -641,6 +643,343 @@ struct type_caster<Type, enable_if_t<is_eigen_other<Type>::value>> {
     using cast_op_type = Type;
 };
 
+template <int Options>
+struct eigen_to_numpy {};
+
+template <>
+struct eigen_to_numpy<Eigen::ColMajor> {
+    static constexpr int flag = array::f_style;
+};
+
+template <>
+struct eigen_to_numpy<Eigen::RowMajor> {
+    static constexpr int flag = array::c_style;
+};
+
+template <class T>
+struct eigen_helper {};
+
+template <typename E, int dim, int O>
+struct eigen_helper<Eigen::Tensor<E, dim, O>> {
+    using Type = Eigen::Tensor<E, dim, O>;
+    using ConstType = Eigen::Tensor<const E, dim, O>;
+
+    using Element = E;
+
+    static constexpr int N = dim;
+    static constexpr int Options = O;
+
+    using ValidType = void;
+
+    static std::array<Eigen::Index, N> get_shape(const Type &f) { return f.dimensions(); }
+
+    static constexpr bool is_correct_shape(const std::array<Eigen::Index, N> & /*shape*/) {
+        return true;
+    }
+};
+
+template <typename E, typename std::ptrdiff_t... Indices, int O>
+struct eigen_helper<Eigen::TensorFixedSize<E, Eigen::Sizes<Indices...>, O>> {
+    using Type = Eigen::TensorFixedSize<E, Eigen::Sizes<Indices...>, O>;
+    using ConstType = Eigen::TensorFixedSize<const E, Eigen::Sizes<Indices...>, O>;
+
+    using Element = E;
+    static constexpr int N = Eigen::Sizes<Indices...>::count;
+    static constexpr int Options = O;
+
+    using ValidType = void;
+
+    static std::array<Eigen::Index, N> get_shape(const Type & /*f*/) { return get_shape(); }
+
+    static constexpr std::array<Eigen::Index, N> get_shape() { return {{Indices...}}; }
+
+    static bool is_correct_shape(const std::array<Eigen::Index, N> &shape) {
+        return get_shape() == shape;
+    }
+};
+
+template <typename Type>
+struct type_caster<Type, typename eigen_helper<Type>::ValidType> {
+    using H = eigen_helper<Type>;
+    PYBIND11_TYPE_CASTER(Type, const_name("eigen::Tensor"));
```

Comment:
This is consistent, the type_caster should inform the python user what the python type of the caster looks like (presumable and ndarray)?

Reply:
Done